connector/src/main/scala/com/datastax/spark/connector/datasource/CassandraCatalog.scala

```diff
@@ -0,0 +1,472 @@
+import org.apache.spark.sql.connector.expressions.Transform
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.util.{Failure, Success, Try}
+
+class CassandraCatalogException(msg: String) extends IllegalArgumentException(msg)
+
+/**
+  * A Spark Sql Catalog for inter-operation with Cassandra
+  *
+  * Namespaces naturally map to C* Keyspaces, but they are always only a single
+  * element deep.
+  */
+class CassandraCatalog extends CatalogPlugin
+  with TableCatalog
+  with SupportsNamespaces
+  with Logging {
+
+  import CassandraCatalog._
+
+  lazy val sparkSession = SparkSession.active
+
+  private val ReplicationClass = "class"
+  private val ReplicationFactor = "replication_factor"
+  private val DurableWrites = "durable_writes"
+  private val NetworkTopologyStrategy = "networktopologystrategy"
+  private val SimpleStrategy = "simplestrategy"
+
+  /*
+    * Using an exclusion list here because CQL Syntax uses all other options for a NetworkToplogyStrategy
+    * keyspace to represent Datacenter names. We cannot enumerate all valid datacenter names so we have
+    * to forbid options reserved and passed directly by Spark
+    */
+  private val IgnoredReplicationOptions = Seq(
+    ReplicationClass,
+    DurableWrites,
+    SupportsNamespaces.PROP_OWNER,
+    SupportsNamespaces.PROP_LOCATION,
+    SupportsNamespaces.PROP_COMMENT)
+
+  private val PartitionKey = "partition_key"
+  private val ClusteringKey = "clustering_key"
+
+  /*
+  This is accessing a driver internal class, but I feel like it's safer than forbidding Spark Specific properties.
+  Hopefully this will also automatically then allow us to pass new C* compatible options by just updating the driver.
+   */
+  private val CassandraProperties = RelationParser.OPTION_CODECS.keySet().asScala
+
+  var connector: CassandraConnector = _
+  var consolidatedConf: SparkConf = _
+  var catalogOptions: CaseInsensitiveStringMap = _
+  //Something to distinguish this Catalog from others with different hosts
+  var catalogName: String = _
+  var nameIdentifier: String = _
+
+  override def initialize(name: String, options: CaseInsensitiveStringMap): Unit = {
+    catalogOptions = options
+    val sparkConf = sparkSession.sparkContext.getConf
+    consolidatedConf = consolidateConfs(sparkConf, sparkSession.conf.getAll, name, userOptions = options.asScala.toMap)
+    connector = CassandraConnector(consolidatedConf)
+    catalogName = name
+    nameIdentifier = connector.conf.contactInfo.endPointStr()
+  }
+
+  override def name(): String = s"Catalog $catalogName For Cassandra Cluster At $nameIdentifier " //TODO add identifier here
+
+  override def listNamespaces(): Array[Array[String]] = {
+    getMetadata(connector)
+      .getKeyspaces
+      .asScala
+      .map { case (name, _) => Array(name.asInternal()) }
+      .toArray
```

Comment:
Semi-nit: you get a bit nicer symmetry with listTables below via something like:

```scala
  override def listNamespaces(): Array[Array[String]] = {
    getMetadata(connector)
      .getKeyspaces
      .asScala
      .keys
      .map(ident => Array(DriverUtil.toName(ident)))
      .toArray
  }
```

You could also do something like the following which would allow you to pull out the map => Strings of the CqlIdentifiers in it's keys logic found in both functions:

```scala
  override def listNamespaces(): Array[Array[String]] = {
    getMetadata(connector)
      .getKeyspaces
      .asScala
      .keys
      .map(DriverUtil.toName(_))
      .map(Array(_))
      .toArray
  }
```

In theory this application lends itself to functional composition but a few quick attempts to make that work kept causing IntelliJ to yell at me... so I moved on. :)

Reply:
Simplified like the first example, I"m not super worried about getting the functional composition here since we are using the Java Driver's classes anyway.